### PR TITLE
Add support for ROS 2 build types via separate debian templates.

### PR DESCRIPTION
**Pull request by nuclearsandwich**
This PR refactors bloom's debian generator to use different template sources for each different supported build type:
- catkin
- cmake
- ament_cmake
- ament_python

When building out support for ROS 2 in bloom I started pushing the templated debian files beyond what I think is really reasonable. In my opinion the templates should be filling in details that are package specific but in order to achieve support for python in particular I had to use template variables to replace the entire behavior of the debian/rules file.

There are downsides to this approach as well. The largest of them is the duplication of files that don't have any build type specific changes. I don't think factoring templates into snippets is ideal since that's really just trading one kind of template complexity for another. I could probably be convinced otherwise if that's what it takes to use templates rather than variables to vary build types. One thing that I am considering but didn't want to spend time implementing up front is keeping one template directory but suffixing any file with a build type specific difference with the build type. For example, having parallel `rules.em.ament_python` and `rules.em.catkin` files and adding the logic to detect and place only the appropriate files. That minimizes the duplication at the cost of complicating the internal template placement implementation and making the template source folder a bit harder to navigate compared to one directory per build type.

Only one of this or #434 should be merged.

### CODE REVIEW

File: bloom/generators/debian/generator.py [hunk 1]
```diff
@@ -326,6 +328,34 @@ def generate_substitutions_from_package(
...
+            from configparser import SafeConfigParser
+            setup_cfg = SafeConfigParser()
+            setup_cfg.read([setup_cfg_path])
+            if (
+                    setup_cfg.has_option('install', 'install-scripts') or
+                    setup_cfg.has_option('install', 'install_scripts')
+            ):
+                data['pass_install_scripts'] = False
+    else:
+        error('Build type `{}` is not supported by this version of bloom.'.format(build_type),
```
Comment (wjwwood) [inline bloom/generators/debian/generator.py]: Same nitpick as the other pr.

**Comment by wjwwood** [comment]
@nuclearsandwich I prefer this one. It seems like a cleaner approach, and I don't have a strong inclination to introduce the most minimal possible change at the expense of a more reasonable layout, as I think is the trade-off made in https://github.com/ros-infrastructure/bloom/pull/434.

**Comment by nuclearsandwich** [comment]
I've run the current master version of bloom and the version in this branch to compare the generated release assets for a ROS catkin package ([random_numbers](https://github.com/ros-gbp/random_numbers-release)) and a ROS cmake package ([ompl](https://github.com/ros-gbp/ompl-release)).

In each diff you'll see that options set in the `DH_OPTIONS` environment variable have been moved to the `dh` invocation line per the recommendation of `debhelper (7)`.

>  When using dh(1), it can be passed options that will be passed on to each debhelper command, which is generally better than using DH_OPTIONS

This change was mandatory for ROS 2 because debhelper builds relying on pybuild were not picking up the buildsystem from the environment variable. For all build types it resolves the `Unknown option: buildsystem` warnings during builds on the buildfarm: an arbitrarily picked example from a [lunar build of geometry2](http://build.ros.org/view/Lbin_uX64/job/Lbin_uX64__geometry2__ubuntu_xenial_amd64__binary/10/consoleFull#console-section-15) can be found in [this gist](https://gist.github.com/nuclearsandwich/4dd69b45b55fd8418f519d72e1bc2ff1).

In the cmake diff the `CATKIN_BUILD_BINARY_PACKAGE` config definition is no longer set since I think it's unnecessary for a non-catkin package, even one that depends on catkin for the environment setup scripts and it was warning when present during ROS 2 beta releasing. I also do not set the CMAKE_PREFIX_PATH as there's an environment variable for it in the setup scripts. If we do want to continue setting it explicitly for CMake packages, we should decide if we want to set it to exclusively the installation prefix or if we want to use whatever the value is in the setup scripts (currently just the installation prefix for ROS kinetic and lunar).

### ROS catkin package random_numbers: diff in placed template

```diff
diff -u c-random_numbers-release/debian/rules.em n-random_numbers-release/debian/rules.em
--- c-random_numbers-release/debian/rules.em    2017-10-10 14:39:42.953044303 -0400
+++ n-random_numbers-release/debian/rules.em    2017-10-10 14:39:37.981031272 -0400
@@ -8,7 +8,6 @@
 
 # Uncomment this to turn on verbose mode.
 export DH_VERBOSE=1
-export DH_OPTIONS=-v --buildsystem=cmake
 # TODO: remove the LDFLAGS override.  It's here to avoid esoteric problems
 # of this sort:
 #  https://code.ros.org/trac/ros/ticket/2977
@@ -20,7 +19,7 @@
 export DEB_CXXFLAGS_MAINT_APPEND=-DNDEBUG
 
 %:
-       dh  $@@
+       dh $@@ -v --buildsystem=cmake
 
 override_dh_auto_configure:
        # In case we're installing to a non-standard location, look for a setup.sh
Common subdirectories: c-random_numbers-release/debian/source and n-random_numbers-release/debian/source
```

### ROS catkin package random_numbers: diff in generated debian files

```diff
diff -u c-random_numbers-release/debian/rules n-random_numbers-release/debian/rules
--- c-random_numbers-release/debian/rules       2017-10-10 14:45:38.621968794 -0400
+++ n-random_numbers-release/debian/rules       2017-10-10 14:45:35.573960922 -0400
@@ -8,7 +8,6 @@
 
 # Uncomment this to turn on verbose mode.
 export DH_VERBOSE=1
-export DH_OPTIONS=-v --buildsystem=cmake
 # TODO: remove the LDFLAGS override.  It's here to avoid esoteric problems
 # of this sort:
 #  https://code.ros.org/trac/ros/ticket/2977
@@ -20,7 +19,7 @@
 export DEB_CXXFLAGS_MAINT_APPEND=-DNDEBUG
 
 %:
-       dh  $@
+       dh $@ -v --buildsystem=cmake
 
 override_dh_auto_configure:
        # In case we're installing to a non-standard location, look for a setup.sh
Common subdirectories: c-random_numbers-release/debian/source and n-random_numbers-release/debian/source
```

### ROS cmake package ompl: diff in placed template files

```diff
diff -u c-ompl-release/debian/rules.em n-ompl-release/debian/rules.em
--- c-ompl-release/debian/rules.em      2017-10-10 15:16:46.846741798 -0400
+++ n-ompl-release/debian/rules.em      2017-10-10 15:16:41.670728621 -0400
@@ -8,7 +8,6 @@
 
 # Uncomment this to turn on verbose mode.
 export DH_VERBOSE=1
-export DH_OPTIONS=-v --buildsystem=cmake
 # TODO: remove the LDFLAGS override.  It's here to avoid esoteric problems
 # of this sort:
 #  https://code.ros.org/trac/ros/ticket/2977
@@ -20,17 +19,14 @@
 export DEB_CXXFLAGS_MAINT_APPEND=-DNDEBUG
 
 %:
-       dh  $@@
+       dh $@@ -v --buildsystem=cmake
 
 override_dh_auto_configure:
        # In case we're installing to a non-standard location, look for a setup.sh
        # in the install tree that was dropped by catkin, and source it.  It will
        # set things like CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
        if [ -f "@(InstallationPrefix)/setup.sh" ]; then . "@(InstallationPrefix)/setup.sh"; fi && \
-       dh_auto_configure -- \
-               -DCATKIN_BUILD_BINARY_PACKAGE="1" \
-               -DCMAKE_INSTALL_PREFIX="@(InstallationPrefix)" \
-               -DCMAKE_PREFIX_PATH="@(InstallationPrefix)"
+       dh_auto_configure -- -DCMAKE_INSTALL_PREFIX="@(InstallationPrefix)"
 
 override_dh_auto_build:
        # In case we're installing to a non-standard location, look for a setup.sh
Common subdirectories: c-ompl-release/debian/source and n-ompl-release/debian/source

```

### ROS cmake package ompl: diff in generated debian files

```diff
diff -u c-ompl-release/debian/rules n-ompl-release/debian/rules
--- c-ompl-release/debian/rules 2017-10-10 15:29:45.752724322 -0400
+++ n-ompl-release/debian/rules 2017-10-10 15:29:42.100715027 -0400
@@ -8,7 +8,6 @@
 
 # Uncomment this to turn on verbose mode.
 export DH_VERBOSE=1
-export DH_OPTIONS=-v --buildsystem=cmake
 # TODO: remove the LDFLAGS override.  It's here to avoid esoteric problems
 # of this sort:
 #  https://code.ros.org/trac/ros/ticket/2977
@@ -20,17 +19,14 @@
 export DEB_CXXFLAGS_MAINT_APPEND=-DNDEBUG
 
 %:
-       dh  $@
+       dh $@ -v --buildsystem=cmake
 
 override_dh_auto_configure:
        # In case we're installing to a non-standard location, look for a setup.sh
        # in the install tree that was dropped by catkin, and source it.  It will
        # set things like CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
        if [ -f "/opt/ros/kinetic/setup.sh" ]; then . "/opt/ros/kinetic/setup.sh"; fi && \
-       dh_auto_configure -- \
-               -DCATKIN_BUILD_BINARY_PACKAGE="1" \
-               -DCMAKE_INSTALL_PREFIX="/opt/ros/kinetic" \
-               -DCMAKE_PREFIX_PATH="/opt/ros/kinetic"
+       dh_auto_configure -- -DCMAKE_INSTALL_PREFIX="/opt/ros/kinetic"
 
 override_dh_auto_build:
        # In case we're installing to a non-standard location, look for a setup.sh
Common subdirectories: c-ompl-release/debian/source and n-ompl-release/debian/source
```

**Comment by nuclearsandwich** [comment]
@wjwwood is there a current rpm support maintainer / ombudsman for Bloom? I wanted to replicate the new structure for RPMs but it doesn't look like we exercise it under tests so a set of dedicated eyes would be good.

**Comment by wjwwood** [comment]
@nuclearsandwich no, I took the contribution so that it could be used by the team of guys helping support fedora at the time. I don't have any special insight or expertise in that department.

**Comment by wjwwood** [comment]
@cottsay may have interest/time to have a look, but I don't know how busy he is.

**Comment by nuclearsandwich** [comment]
Sounds good. I want to make a documentation pass for the new build type changes and if @cottsay or any other rpm folks have the time to review I'll do my best to incorporate feedback.

Otherwise I think this is ready as soon as I get some docs out. There are no intentional user-facing breaking changes for catkin and cmake builds so I do not believe any user documentation needs to be updated except in the ROS 2 umbrella which will be supported by the next release of Bloom after this merges.

**Comment by cottsay** [comment]
I should have spare cycles to review changes from the RPM side.

I'm a little out of the loop on ROS2 stuff, so you might have to explain things a little bit. If I understand this correctly, we just want to (conditionally) use different varieties of build steps, whereas right now the RPM always uses a CMake call?

**Comment by nuclearsandwich** [comment]
> I should have spare cycles to review changes from the RPM side.

Thanks, I very much appreciate it.

> we just want to (conditionally) use different varieties of build steps, whereas right now the RPM always uses a CMake call?

That's it, essentially. ament, catkin, and generic CMake builds all still run mostly the same but with different sets of build arguments. The biggest new change which ultimately prompted breaking the different build types up into different templates instead of using a whole bunch of template variables was python setuptools support for ament.

Where there used to be one set of templates per type of generated package there is now one per generated package per build type. 

I haven't implemented the new ROS 2 build types for the rpm generator yet, just set up the same new template structure for catkin and cmake packages and added [explicit failure](https://github.com/ros-infrastructure/bloom/pull/435/files#diff-befe9ce877a3c0fd439fe95ae71e14b9R254) when trying to generate an rpm of the new as yet unsupported build types. I don't think we need complete rpm support right now, and I'm certainly not asking you to implement it for this PR. But I would like to be reasonably confident that I didn't introduce any regressions into the RPM generation.

**Comment by nuclearsandwich** [comment]
Added some very light developer documentation for the new feature [(rendered)](https://github.com/ros-infrastructure/bloom/blob/a40d8f831827dd24da3ba095d7e90a777a0cfd3c/bloom/generators/README.md) over time hopefully we can flesh that out with info about platform support, rosdep, and the like and improve the API and code sharing for generators but it'll do for now.

Modulo any feedback on the added paragraph of docs I think this is ready to merge and release.

**Comment by wjwwood** [comment]
> Modulo any feedback on the added paragraph of docs I think this is ready to merge and release.

Sounds good to me. I'll let you merge it, and I should show you how to do a release in case you ever need to do it without me. (we can discuss off GitHub)

**Pull request merged**